docs/notebooks/How_JAX_primitives_work.ipynb

```diff
@@ -0,0 +1,1638 @@
+        "with expectNotImplementedError():\n",
+        "  api.vmap(sq_add_prim, in_axes=0, out_axes=0)(onp.array([2., 3.]),\n",
+        "                                               onp.array([10., 20.]))"
+      ],
+      "execution_count": 0,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "call sq_add_prim(Traced<ShapedArray(float32[])>, Traced<ShapedArray(float32[])>)\n",
+            "  call ma_prim(Traced<ShapedArray(float32[])>, Traced<ShapedArray(float32[])>, Traced<ShapedArray(float32[])>)\n",
+            "\n",
+            "Found expected exception:\n"
+          ],
+          "name": "stdout"
+        },
+        {
+          "output_type": "stream",
+          "text": [
+            "Traceback (most recent call last):\n",
+            "  File \"/usr/local/lib/python3.6/dist-packages/jax/interpreters/batching.py\", line 162, in get_primitive_batcher\n",
+            "    return primitive_batchers[p]\n",
+            "KeyError: ma\n",
+            "\n",
+            "During handling of the above exception, another exception occurred:\n",
+            "\n",
+            "Traceback (most recent call last):\n",
+            "  File \"<ipython-input-120-e1a8bde26b60>\", line 3, in <module>\n",
+            "    onp.array([10., 20.]))\n",
+            "  File \"/usr/local/lib/python3.6/dist-packages/jax/api.py\", line 609, in batched_fun\n",
+            "    lambda: _flatten_axes(out_tree(), out_axes))\n",
+            "  File \"/usr/local/lib/python3.6/dist-packages/jax/interpreters/batching.py\", line 41, in batch\n",
+            "    out_vals, out_dims = batch_fun(fun, in_vals, in_dims)\n",
+            "NotImplementedError: Batching rule for 'ma' not implemented\n"
+          ],
+          "name": "stderr"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "gILasMiP6elR",
+        "colab_type": "text"
+      },
+      "source": [
+        "We need to tell JAX how to evaluate the batched version of the primitive. In this particular case, the `ma_prim` already operates pointwise for any dimension of input vectors. So the"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "KQfeqRIrP7zg",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "from jax import batching\n",
+        "\n",
+        "\n",
+        "@trace(\"ma_batch\")\n",
+        "def ma_batch(vector_arg_values, batch_axes):\n",
+        "  \"\"\"Computes the batched version of the primitive.\n",
+        "  \n",
+        "  This must be a JAX-traceable function.\n",
+        "  \n",
+        "  Since the ma primitive already operates on tensors, to batch it we use\n",
+        "  the primitive itself. This works as long as the input are batched along\n",
+        "  the same axes. The result is batched along the axis that the inputs\n",
```

Comment:
Do we need to worry about broadcasting here? What if the inputs have different numbers of dimensions?

Reply:
I added a bit of clarification that this works when the inputs have the same dimensions. Do you think we need to do something more elaborate for the purposes of this doc?